### PR TITLE
Add contract-negotiation management endpoints

### LINKED ISSUE #589: DataManagementApi: Create API for ContractNegotiations
_This issue only deals with the API, i.e. a stub for the `*Controller` class. An implementation is explicitly NOT to be done._

## API Endpoints
- `GET /api/v1/contractnegotiations`: gets all ContractNegotiations, accepts paging spec, returns `List<ContractNegotiationDto>`
- `GET /api/v1/contractnegotiations/{id}`: gets one particular `ContractNegotiationDto`
- `GET /api/v1/contractnegotiations/{id}/state`: gets the state of one particular negotiation, useful for polling 
- `POST /api/v1/contractnegotiations/{id}/cancel`: moves the ContractNegotation to the `CANCELLING/CANCELLED` state. Throws `IllegalStateException` if not possible. Returns `void`
- `POST /api/v1/contractnegotiations/{id}/decline`: moves the ContractNegotation to the `DECLINING/DECLINED` state. Throws `IllegalStateException` if not possible. Returns `void`.

## DTOs
```java
public class ContractNegotiationDto{
    private String id;
    private String counterPartyAddress;
    private String protocol;
    private Type type = Type.CONSUMER;
    private String state;
    private String errorDetail;
    private String contractAgreementId; // is null until state == CONFIRMED
    //getters, setters,...
}
```
Code should go into `extensions:api:data-management:contract-negotiation`

### PR DESCRIPTION
**Pull request by florianrusch-zf**
closes #589

### CODE REVIEW

**Comment by florianrusch-cluetec** [comment]
@paullatzelsperger If you want, you can start reviewing this. But it should be merged after your #641

**Comment by paullatzelsperger** [comment]
> @paullatzelsperger If you want, you can start reviewing this. But it should be merged after your #641

sure, once you're ready to review, just mark the PR as such and I'll look at it.

**Comment by juliapampus** [comment]
@florianrusch-cluetec There are already some REST endpoints for negotiation management in the `ClientController` class. Can you please take a look what is duplicated and what may be missing in your implementation, and remove the corresponding methods from the `ClientController`?

On top of that: Is it currently necessary to create this package structurer? `api-core` and `control` may not be the best solution but naming it `api.datamanagement.contractnegotiation` is maybe a little bit too long.

**Comment by florianrusch-cluetec** [comment]
> @florianrusch-cluetec There are already some REST endpoints for negotiation management in the `ClientController` class. Can you please take a look what is duplicated and what may be missing in your implementation, and remove the corresponding methods from the `ClientController`?

I thought, that this would be part of the issue which fills this endpoints with logic. Else we would remove the possibility to access this functionality in the meantime.

> On top of that: Is it currently necessary to create this package structurer? `api-core` and `control` may not be the best solution but naming it `api.datamanagement.contractnegotiation` is maybe a little bit too long.

We try to group all data management endpoints in one place. Therefore all endpoints will be submodules of the `extensions:api:datamanagement` module. Please have a look at #542

File: extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/ContractNegotiationController.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import org.eclipse.dataspaceconnector.spi.query.SortOrder;
+
+import java.util.Collections;
+import java.util.List;
+
+import static java.lang.String.format;
+
+@Consumes({ MediaType.APPLICATION_JSON })
+@Produces({ MediaType.APPLICATION_JSON })
+@Path("/contractnegotitations")
```
[the remark this answers is not in the document]
Comment (florianrusch-cluetec) [reply extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/ContractNegotiationController.java:36]: Good catch!

File: extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDto.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
+import org.eclipse.dataspaceconnector.spi.types.domain.contract.negotiation.ContractNegotiation.Type;
+
+@JsonDeserialize(builder = ContractNegotiationDto.Builder.class)
+public class ContractNegotiationDto {
+    private String id;
+    private String counterPartyAddress;
+    private String protocol;
```
Comment (paullatzelsperger) [inline extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDto.java:25]: should be "ids-multipart" by default. Either initialize in constructor or as initializer.

File: extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDto.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        }
+
+        public Builder contractAgreementId(String contractAgreementId) {
+            dto.contractAgreementId = contractAgreementId;
+            return this;
+        }
+
+        public Builder type(Type type) {
+            if (type != Type.CONSUMER) {
+                throw new IllegalArgumentException("The type can only be CONSUMER");
```
Comment (paullatzelsperger) [inline extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDto.java:107]: DTOs should not validate input, especially not in builder methods.

File: extensions/api/data-management/contractnegotiation/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDtoTest.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        var dto = ContractNegotiationDto.Builder.newInstance()
+                .contractAgreementId("test-contract-agreement-id")
+                .counterPartyAddress("test-counter-party-address")
+                .errorDetail("test-error-detail")
+                .protocol("test-protocol")
+                .state("test-state")
+                .id("test-id")
+                .build();
+
+        var str = objectMapper.writeValueAsString(dto)
```
Comment (paullatzelsperger) [inline extensions/api/data-management/contractnegotiation/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDtoTest.java:65]: Check comment in `ContractNegotiationDto`.

File: extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/ContractNegotiationController.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                                                        @QueryParam("sortField") String sortField) {
+        var spec = QuerySpec.Builder.newInstance()
+                .offset(offset)
+                .limit(limit)
+                .sortField(sortField)
+                .filter(filterExpression)
+                .sortOrder(sortOrder)
+                .build();
+
+        monitor.debug(format("Get all contract definitions %s", spec));
```
Comment (paullatzelsperger) [inline extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/ContractNegotiationController.java:59]: FYI: We'll have to remove those debug outputs once the implementation goes in.

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]
Comments in code. 
Unless i have wrong information, the Copyright should belong to ZF, right?

File: samples/04.0-file-transfer/consumer/build.gradle.kts [hunk 6]
```diff
@@ -27,7 +27,6 @@ dependencies {
 
     implementation(project(":extensions:in-memory:assetindex-memory"))
     implementation(project(":extensions:in-memory:transfer-store-memory"))
-    implementation(project(":extensions:in-memory:assetindex-memory"))
     implementation(project(":extensions:in-memory:negotiation-store-memory"))
```
[the remark this answers is not in the document]
Comment (florianrusch-cluetec) [reply samples/04.0-file-transfer/consumer/build.gradle.kts:30]: Take a look two lines above ;-) It was something I identified while trying the sample in the onboarding progress and thought a own PR for this would be an overkill

[the remark this answers is not in the document]
Comment (paullatzelsperger) [reply samples/04.0-file-transfer/consumer/build.gradle.kts:30]: Whups, didn't see it in the diff. Fell free to resolve then.

File: extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDto.java (hunk 3 above)
Comment (florianrusch-cluetec) [reply extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDto.java:107]: I'm totally with you! But how should I then handle the type attribute? My first attempt was to not implement the type method in the Builder class because it should be set automatically at new object anyway. But then I got this error from the serialization test:
```
Unrecognized field "type" (class org.eclipse.dataspaceconnector.api.datamanagement.contractnegotiation.model.ContractNegotiationDto$Builder), not marked as ignorable (6 known properties: "state", "contractAgreementId", "protocol", "errorDetail", "counterPartyAddress", "id"])
 at [Source: (String)"{"id":"test-id","counterPartyAddress":"test-counter-party-address","protocol":"test-protocol","type":"CONSUMER","state":"test-state","errorDetail":"test-error-detail","contractAgreementId":"test-contract-agreement-id"}"; line: 1, column: 103] (through reference chain: org.eclipse.dataspaceconnector.api.datamanagement.contractnegotiation.model.ContractNegotiationDto$Builder["type"])
com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "type" (class org.eclipse.dataspaceconnector.api.datamanagement.contractnegotiation.model.ContractNegotiationDto$Builder), not marked as ignorable (6 known properties: "state", "contractAgreementId", "protocol", "errorDetail", "counterPartyAddress", "id"])
 at [Source: (String)"{"id":"test-id","counterPartyAddress":"test-counter-party-address","protocol":"test-protocol","type":"CONSUMER","state":"test-state","errorDetail":"test-error-detail","contractAgreementId":"test-contract-agreement-id"}"; line: 1, column: 103] (through reference chain: org.eclipse.dataspaceconnector.api.datamanagement.contractnegotiation.model.ContractNegotiationDto$Builder["type"])
```

The problem is, that he takes all fields from the json and tries to call methods with the same name on the Builder object -> Kabäm there is no function for the type.

A idea that comes up in my mind while writing this: Should I simply implement this?

```java
        public Builder type(Type type) {
            return this;
        }
```

File: extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDto.java (hunk 2 above)
Comment (florianrusch-cluetec) [reply extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDto.java]: Fixed

File: extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDto.java (hunk 3 above)
Comment (paullatzelsperger) [reply extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDto.java:107]: there _should_ be a method to set the type. the type is _not_ fixed, it simply is initialized with `CONSUMER` by default. 

As a bit of general information: the `@JsonIgnore` annotation is used to exclude fields from serialization/deserialization, but that's _not_ the case here.

File: extensions/api/data-management/contractnegotiation/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/ContractNegotiationApiControllerIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+ */
+
+package org.eclipse.dataspaceconnector.api.datamanagement.contractnegotiation;
+
+import java.io.IOException;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+
+import okhttp3.OkHttpClient;
```
[the remark this answers is not in the document]
Comment (paullatzelsperger) [reply extensions/api/data-management/contractnegotiation/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/ContractNegotiationApiControllerIntegrationTest.java:21]: don't know where your `.editorconfig` comes from, we don't (and probably won't) provide one. please checkout `styleguide.md`, all you need to know is in there.

File: extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDto.java (hunk 3 above)
Comment (florianrusch-cluetec) [reply extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/model/ContractNegotiationDto.java:107]: Ahhh okay - interpreted it as been a fix value. Will fix it

**Comment by florianrusch-zf** [comment]
@ndr-brt  @paullatzelsperger should now be mergeable ;-)

File: resources/openapi/yaml/contractnegotiation.yaml [hunk 8]
```diff
@@ -0,0 +1,122 @@
+openapi: 3.0.1
```
Comment (paullatzelsperger) [inline resources/openapi/yaml/contractnegotiation.yaml:1]: there are two yaml files: `contract-negotiation.yaml` and `contractnegotiaion.yaml`, we only need one.

File: settings.gradle.kts [hunk 9]
```diff
@@ -156,7 +159,6 @@ include(":samples:04.0-file-transfer:consumer")
 include(":samples:04.0-file-transfer:provider")
 include(":samples:04.0-file-transfer:api")
 include(":samples:04.0-file-transfer:transfer-file")
-include(":samples:04.0-file-transfer:integration-tests")
```
Comment (paullatzelsperger) [inline settings.gradle.kts:159]: why was this removed? rebase problem?

File: resources/openapi/yaml/contractnegotiation.yaml (hunk 8 above)
Comment (florianrusch-zf) [reply resources/openapi/yaml/contractnegotiation.yaml:1]: Good catch - is a renaming leftover. Will fix it

File: settings.gradle.kts (hunk 9 above)
Comment (florianrusch-zf) [reply settings.gradle.kts]: yes. I fixed it

**Review by paullatzelsperger — APPROVED** [review]
